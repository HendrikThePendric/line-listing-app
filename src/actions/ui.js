import {
    ADD_UI_LAYOUT_DIMENSIONS,
    SET_UI_LAYOUT,
    SET_UI_OPTION,
    SET_UI_OPTIONS,
} from '../reducers/ui'

export const acSetUiOptions = value => ({
    type: SET_UI_OPTIONS,
    value,
})

export const acSetUiOption = value => ({
    type: SET_UI_OPTION,
    value,
})

export const acAddUiLayoutDimensions = value => ({
    type: ADD_UI_LAYOUT_DIMENSIONS,
    value,
})

export const acSetUiLayout = value => ({
    type: SET_UI_LAYOUT,
    value,
})
