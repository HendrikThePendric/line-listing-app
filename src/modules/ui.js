import {
    getAdaptedUiLayoutByType,
    VIS_TYPE_PIVOT_TABLE,
    VIS_TYPE_PIE,
    layoutGetAxisIdDimensionIdsObject,
    layoutGetDimensionIdItemIdsObject,
} from '@dhis2/analytics'

const defaultUiAdapter = ui => ({
    ...ui,
    layout: getAdaptedUiLayoutByType(ui.layout, VIS_TYPE_PIE), // TODO: Change this once LL has its own type (temp fix since pie has same axes as LL)),
    // FIXME: Note that using VIS_TYPE_PIE means all but 1 items will be moved to filter
})

export const getAdaptedUiByType = ui => {
    switch (ui.type) {
        case VIS_TYPE_PIVOT_TABLE:
            return ui
        default:
            return defaultUiAdapter(ui)
    }
}

export const getUiFromVisualization = (vis, currentState = {}) => ({
    ...currentState,
    // type: vis.type || defaultVisType,
    // options: getOptionsFromVisualization(vis),
    layout: layoutGetAxisIdDimensionIdsObject(vis),
    itemsByDimension: layoutGetDimensionIdItemIdsObject(vis),
    // parentGraphMap:
    //     vis.parentGraphMap ||
    //     getParentGraphMapFromVisualization(vis) ||
    //     currentState.parentGraphMap,
})
