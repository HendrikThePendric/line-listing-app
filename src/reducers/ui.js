/*eslint no-unused-vars: ["error", { "ignoreRestSiblings": true }]*/
import {
    DIMENSION_ID_DATA,
    DIMENSION_ID_ORGUNIT,
    DIMENSION_ID_PERIOD,
} from '@dhis2/analytics'
import { getFilteredLayout } from '../modules/layout'
import { getOptionsForUi } from '../modules/options'

export const SET_UI = 'SET_UI'
export const SET_UI_OPTIONS = 'SET_UI_OPTIONS'
export const SET_UI_OPTION = 'SET_UI_OPTION'
export const ADD_UI_LAYOUT_DIMENSIONS = 'ADD_UI_LAYOUT_DIMENSIONS'
export const SET_UI_LAYOUT = 'SET_UI_LAYOUT'

const DEFAULT_UI = {
    options: getOptionsForUi(),
    layout: {
        columns: [DIMENSION_ID_DATA],
        filters: [DIMENSION_ID_ORGUNIT, DIMENSION_ID_PERIOD],
    },
    itemsByDimension: {
        [DIMENSION_ID_ORGUNIT]: [],
        [DIMENSION_ID_PERIOD]: [],
    },
}

export default (state = DEFAULT_UI, action) => {
    switch (action.type) {
        case SET_UI: {
            return {
                ...action.value,
            }
        }
        case SET_UI_OPTIONS: {
            return {
                ...state,
                options: {
                    ...state.options,
                    ...action.value,
                },
            }
        }
        // action.value: transfer object (dimensionId:axisId) saying what to add where: { ou: 'rows' }
        // Reducer takes care of swapping (retransfer) if dimension already exists in layout
        case ADD_UI_LAYOUT_DIMENSIONS: {
            const transfers = {
                ...action.value,
            }

            let newLayout = state.layout

            // Add dimension ids to destination (axisId === null means remove from layout)
            Object.entries(transfers).forEach(
                ([dimensionId, { axisId, index }]) => {
                    if (newLayout[axisId]) {
                        // Filter out transferred dimension id (remove from source)
                        newLayout = getFilteredLayout(newLayout, [dimensionId])
                        if (index === null || index === undefined) {
                            newLayout[axisId].push(dimensionId)
                        } else {
                            newLayout[axisId].splice(index, 0, dimensionId)
                        }
                    }
                }
            )

            return {
                ...state,
                layout: newLayout,
            }
        }
        case SET_UI_LAYOUT: {
            return {
                ...state,
                layout: {
                    ...action.value,
                },
            }
        }
        default:
            return state
    }
}

// Selectors

export const sGetUi = state => state.ui
export const sGetUiOptions = state => sGetUi(state).options
export const sGetUiOption = () => {} // TODO: items stored here should be flattened and reintegrated into sGetUiOptions (above)
export const sGetUiItems = state => sGetUi(state).itemsByDimension
export const sGetUiLayout = state => sGetUi(state).layout

// Selectors level 2

export const sGetUiItemsByDimension = (state, dimension) =>
    sGetUiItems(state)[dimension] || DEFAULT_UI.itemsByDimension[dimension]
