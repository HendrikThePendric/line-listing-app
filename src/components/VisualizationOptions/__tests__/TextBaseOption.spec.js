import { InputField } from '@dhis2/ui'
import { shallow } from 'enzyme'
import React from 'react'
import { TextBaseOption } from '../Options/TextBaseOption'

describe('DV > Options > TextBaseOption', () => {
    let props
    let shallowTextBaseOption
    let onChange

    const textBaseOption = props => {
        shallowTextBaseOption = shallow(<TextBaseOption {...props} />)

        return shallowTextBaseOption
    }

    beforeEach(() => {
        onChange = jest.fn()

        props = {
            value: 'test',
            type: 'text',
            label: 'Input field',
            option: {
                name: 'input1',
            },

            onChange,
        }

        shallowTextBaseOption = undefined
    })

    it('renders a <InputField />', () => {
        expect(textBaseOption(props).find(InputField)).toHaveLength(1)
    })

    it('sets the type prop to what passed in the prop', () => {
        expect(textBaseOption(props).find(InputField).props().type).toEqual(
            props.type
        )
    })

    it('sets the label prop to what passed in the prop', () => {
        expect(textBaseOption(props).find(InputField).props().label).toEqual(
            props.label
        )
    })

    it('sets the value to what passed in the prop', () => {
        expect(textBaseOption(props).find(InputField).props().value).toEqual(
            props.value
        )
    })

    it('sets the placeholder to what passed in the prop', () => {
        props.placeholder = 'placeholder text'

        expect(
            textBaseOption(props).find(InputField).props().placeholder
        ).toEqual(props.placeholder)
    })

    it('sets the width prop to what passed in the prop', () => {
        props.width = '105px'

        expect(
            textBaseOption(props).find(InputField).props().inputWidth
        ).toEqual(props.width)
    })

    it('should trigger the onChange callback on text change', () => {
        const text = textBaseOption(props).find(InputField)

        text.simulate('change', { value: 'test' })

        expect(onChange).toHaveBeenCalled()
    })
})
