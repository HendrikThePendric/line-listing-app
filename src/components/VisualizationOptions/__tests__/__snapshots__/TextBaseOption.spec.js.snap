// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ER > Options > TextBaseOption renders disabled input with type time 1`] = `
<div>
  <div
    class="tabSectionOption"
  >
    <div
      class="jsx-1141087211 "
      data-test="testing-prefix-input"
    >
      <label
        class="jsx-2718078600 disabled"
        data-test="testing-prefix-input-label"
        for="theTime"
      >
        <span
          class="jsx-2718078600"
        >
          Current time
        </span>
      </label>
      <div
        class="jsx-3111771310 "
        data-test="testing-prefix-input-content"
      >
        <div
          class="jsx-2908043323 "
          data-test="dhis2-uicore-box"
        >
          <div
            class="jsx-3353877153 jsx-2422099811 input"
            data-test="dhis2-uicore-input"
          >
            <input
              class="jsx-3353877153 jsx-2422099811 dense disabled"
              disabled=""
              id="theTime"
              name="theTime"
              placeholder="placeholder text"
              type="time"
              value="10:30"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ER > Options > TextBaseOption renders enabled input with type time 1`] = `
<div>
  <div
    class="tabSectionOption"
  >
    <div
      class="jsx-1141087211 "
      data-test="testing-prefix-input"
    >
      <label
        class="jsx-2718078600 "
        data-test="testing-prefix-input-label"
        for="theTime"
      >
        <span
          class="jsx-2718078600"
        >
          Current time
        </span>
      </label>
      <div
        class="jsx-3111771310 "
        data-test="testing-prefix-input-content"
      >
        <div
          class="jsx-2908043323 "
          data-test="dhis2-uicore-box"
        >
          <div
            class="jsx-3353877153 jsx-2422099811 input"
            data-test="dhis2-uicore-input"
          >
            <input
              class="jsx-3353877153 jsx-2422099811 dense"
              id="theTime"
              name="theTime"
              placeholder="placeholder text"
              type="time"
              value="10:30"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
